src/internal/storage/chunk/gc.go

```diff
@@ -0,0 +1,149 @@
+package chunk
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"strconv"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/kv"
+	"github.com/sirupsen/logrus"
+)
+
+// GarbageCollector removes unused chunks from object storage
+type GarbageCollector struct {
+	s *Storage
+}
+
+// NewGC returns a new garbage collector operating on s
+func NewGC(s *Storage) *GarbageCollector {
+	return &GarbageCollector{s: s}
+}
+
+// RunForever calls RunOnce until the context is cancelled, logging any errors.
+func (gc *GarbageCollector) RunForever(ctx context.Context) error {
+	ticker := time.NewTicker(time.Minute)
+	defer ticker.Stop()
+	for {
+		if err := gc.RunOnce(ctx); err != nil {
+			select {
+			case <-ctx.Done():
+				return err
+			default:
+			}
+			logrus.Errorf("during chunk GC: %v", err)
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+		}
+	}
+}
+
+// RunOnce runs 1 cycle of garbage collection.
+func (gc *GarbageCollector) RunOnce(ctx context.Context) (retErr error) {
+	rows, err := gc.s.db.QueryContext(ctx, `
+	SELECT chunk_id, gen, uploaded FROM storage.chunk_objects
+	WHERE tombstone = true
+	`)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := rows.Close(); retErr == nil {
+			retErr = err
+		}
+	}()
+	for rows.Next() {
+		var ent Entry
+		if err := sqlx.StructScan(rows, &ent); err != nil {
+			return err
+		}
+		if !ent.Uploaded {
+			logrus.Warnf("possibility for untracked chunk %s", chunkPath(ent.ChunkID, ent.Gen))
+		}
+		if err := gc.deleteOne(ctx, ent); err != nil {
+			return err
+		}
+	}
+	return rows.Err()
+}
+
+func (gc *GarbageCollector) deleteOne(ctx context.Context, ent Entry) error {
+	if err := gc.deleteObject(ctx, ent.ChunkID, ent.Gen); err != nil {
+		return err
+	}
+	return gc.deleteEntry(ctx, ent.ChunkID, ent.Gen)
+}
+
+func (gc *GarbageCollector) deleteObject(ctx context.Context, chunkID ID, gen uint64) error {
+	return gc.s.store.Delete(ctx, chunkKey(chunkID, gen))
```

Comment:
I think we could run into issues if we return an error here when the object does not exist. For example, if the machine crashes immediately after this call and before deleting the entry, then I think the next garbage collector would be stuck in a loop once it reached this point again for the object.

Reply:
I think delete is idempotent.  If it's not at the object level; I propose changing it at the kv level.

Reply:
Okay so delete is not idempotent on GCS
https://cloud.google.com/storage/docs/xml-api/delete-object

But it is in AWS.
https://docs.aws.amazon.com/sdkfornet1/latest/apidocs/html/M_Amazon_S3_AmazonS3Client_DeleteObject.htm

So I changed my mind, this needs to be uniform before it leaves the obj package.

Reply:
Okay, so are you thinking we should have a standard error for this case or just requiring that the clients don't return an error if the object does not exist?

Reply:
I like no error. I think that is more useful generally. I can't think of a use case where we care if there was something there.

Reply:
Okay, for the sake of this P.R. lets assume this is the behavior. And track it here:
https://github.com/pachyderm/pachyderm/pull/5866#issuecomment-804299195